compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -120,11 +137,31 @@ impl<'context> Elaborator<'context> {
                 );
             }
         }
+
+        // Check that any types with a type kind match the expected type kind supplied to this function
+        if let Type::NamedGeneric(_, name, resolved_kind) = &resolved_type {
+            if matches!(resolved_kind, TypeKind::Numeric) && matches!(kind, TypeKind::Normal) {
```

Comment:
This shouldn't be specific to named generics, it should apply to every type:
```suggestion
        if resolved_type.kind() != kind
```

Reply:
did it manually as I was not able to commit the suggestion after changes to the diff

Reply:
Ended up going back to de-structuring `resolved_type` so that I could clearly fetch the name in the case of NamedGeneric. But I replaced the matches with `resolved_kind != kind`